### PR TITLE
bind wlr_text_input_v3.h

### PR DESCRIPTION
**Pull request by novakne**

### CODE REVIEW

File: src/types/text_imput_v3.zig [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        width: i32,
+        height: i32,
+    };
+
+    text_change_cause: u32,
+    features: u32,
+
+    seat: *wlr.Seat,
+    ressource: *wl.Ressource,
+    focused_surface: *wlr.Surface,
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:56]: This one needs to be optional. In general, we need to go through most stored pointers to check if they are ever set to null.

File: src/types/text_imput_v3.zig [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    events: extern struct {
+        enable: wl.Signal(*wlr.TextInputV3),
+        commit: wl.Signal(*wlr.TextInputV3),
+        disable: wl.Signal(*wlr.TextInputV3),
+        enable: wl.Signal(*wlr.TextInputV3),
+    },
+};
+
+extern fn wlr_text_input_v3_send_enter(text_input: *wlr.TextInputV3, surface: *wlr.Surface) void;
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:77]: This and the other `wlr_text_input_v3` functions should go inside the `TextInputV3` namespace.

**Review by ifreund — CHANGES_REQUESTED** [review]

File: src/types/text_imput_v3.zig [hunk 3]
```diff
@@ -0,0 +1,114 @@
+const wlr = @import("../wlroots.zig");
+
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+
+pub const TextInputV3Manager = extern struct {
+    global: *wl.Global,
+    text_inputs: wl.list.Head(TextInputV3Manager, "link"),
+
+    display_destroy: wl.Listener(*wl.Display),
```
[the remark this answers is not in the document]
Comment (novakne) [reply src/types/text_imput_v3.zig:10]: Oh that's a difference with the main wlroots right? I think I read about it but forget about it after

File: src/types/text_imput_v3.zig [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+pub const TextInputV3 = extern struct {
+    // Not sure about this
+    pub const Features = extern struct {
+        pub const surrounding_text = 1 << 0;
+        pub const content_type = 1 << 1;
+        pub const cursor_rectangle = 1 << 2;
+    };
+
+    pub const State = extern struct {
```
[the remark this answers is not in the document]
Comment (novakne) [reply src/types/text_imput_v3.zig:31]: Huh my brain died here or what.. sorry

[the remark this answers is not in the document]
Comment (novakne) [reply src/types/text_imput_v3.zig:31]: oh wait they are here but I didn't close the State enum

**Comment by novakne** [comment]
Well that's was a lot of errors, should be good now

File: src/types/text_imput_v3.zig [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        pub const Surrounding = extern struct {
+            text: ?[*:0]u8,
+            cursor: u32,
+            anchor: u32,
+        };
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:35]: This is a field, as is content_type and cursor_rectangle.
```suggestion
        surrounding: extern struct {
            text: ?[*:0]u8,
            cursor: u32,
            anchor: u32,
        },
```

File: src/types/text_imput_v3.zig [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        before_length: u32,
+        after_length: u32,
+    ) void {
+        return wlr_text_input_v3_send_delete_surrounding_text(text_input, before_length, after_length);
+    }
+
+    extern fn wlr_text_input_v3_send_done(text_input: *wlr.TextInputV3) void;
+    pub fn sendDone(text_input: *wlr.TextInputV3) void {
+        return wlr_text_input_v3_send_done(text_input);
+    }
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:112]: there's no need for a wrapper here, just do 
```zig
pub const sendDone = wlr_text_input_v3_send_done;
```
The same goes for the other functions as well.

File: src/types/text_imput_v3.zig [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    },
+
+    extern fn wlr_text_input_v3_create(server: *wl.Server) ?*wlr.TextInputV3Manager;
+    pub fn create(server: *wl.Server) !*wlr.TextInputV3Manager {
+        return wlr_text_input_v3_create(server) orelse error.OutOfMemory;
+    }
+};
+
+pub const TextInputV3 = extern struct {
+    pub const Features = extern struct {
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:24]: This struct has no fields, therefore it is a namespace and snake_case. The extern is also unnecessary.
```suggestion
    pub const features = struct {
```

**Review by ifreund — CHANGES_REQUESTED** [review]

File: src/types/text_imput_v3.zig [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+pub const TextInputV3 = extern struct {
+    pub const features = struct {
+        pub const surrounding_text = 1 << 0;
+        pub const content_type = 1 << 1;
+        pub const cursor_rectangle = 1 << 2;
+    };
+
+    pub const State = extern struct {
```
Comment (ifreund) [inline src/types/text_imput_v3.zig:30]: The fields of this struct are in the wrong order.

**Comment by novakne** [comment]
damn, sorry for all this work on the review

**Review by ifreund — CHANGES_REQUESTED** [review]
I don't see any other issues, so with these last two fixes this should be good to merge.

**Review by ifreund — APPROVED** [review]
LGTM!

**Comment by novakne** [comment]
Thanks for your patience, next time will be better :)

**Pull request merged**